File: src/build-info.ts
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";

/**
 * Build metadata stamped into the packaged macOS app by
 * `scripts/build-macos-app.sh`. Only `buildId` is load-bearing (it gates the
 * app's version-skew daemon restart); the rest is diagnostic.
 */
export interface BuildInfo {
	buildId: string;
	commit?: string;
	version?: string;
	builtAt?: string;
}

/**
 * Read the `buildId` the build script stamped next to the bundled CLI. The file
 * lives at `<cli-dir>/build-info.json`, one level up from this module's `src/`
 * dir, so the default location resolves from `import.meta.dir`.
 *
 * Returns `null` when the file is absent or unparseable — the dev / source-tree
 * case (no build step ran), where the app intentionally suppresses the
 * version-skew auto-restart rather than thrash a daemon it cannot compare.
 */
export function readBuildId(
	buildInfoDir: string = join(import.meta.dir, ".."),
): string | null {
	const path = join(buildInfoDir, "build-info.json");
	if (!existsSync(path)) return null;
	try {
		const parsed: unknown = JSON.parse(readFileSync(path, "utf8"));
		if (
			parsed !== null &&
			typeof parsed === "object" &&
			"buildId" in parsed &&
			typeof (parsed as { buildId: unknown }).buildId === "string" &&
			(parsed as { buildId: string }).buildId.length > 0
		) {
			return (parsed as { buildId: string }).buildId;
		}
		return null;
	} catch {
		return null;
	}
}

File: macos/AgentVoiceApp/Sources/AgentVoiceCore/BuildInfo.swift
import Foundation

/// Build metadata stamped into the packaged app by `scripts/build-macos-app.sh`
/// at `Contents/Resources/agent-voice/build-info.json`. Only `buildId` is
/// load-bearing — it gates the version-skew daemon restart; the rest is
/// diagnostic. Mirrors the TypeScript `BuildInfo` in `src/build-info.ts`.
public struct BuildInfo: Codable, Equatable, Sendable {
    public let buildId: String
    public let commit: String?
    public let version: String?
    public let builtAt: String?

    public init(buildId: String, commit: String? = nil, version: String? = nil, builtAt: String? = nil) {
        self.buildId = buildId
        self.commit = commit
        self.version = version
        self.builtAt = builtAt
    }

    /// Read the `buildId` from `build-info.json` in the given directory, or `nil`
    /// when the file is absent, unreadable, unparseable, or carries an empty id —
    /// the dev / source-tree case, where the app suppresses the auto-restart
    /// rather than act on a build id it cannot trust.
    public static func readBuildId(fromDirectory directory: URL) -> String? {
        let url = directory.appendingPathComponent("build-info.json")
        guard let data = try? Data(contentsOf: url),
              let info = try? JSONDecoder().decode(BuildInfo.self, from: data),
              !info.buildId.isEmpty
        else {
            return nil
        }
        return info.buildId
    }
}
